mitmproxy/proxy/protocol2/websocket.py

```diff
@@ -16,14 +18,64 @@ def __init__(self, context: ClientServerContext, handshake_flow: http.HTTPFlow):
         super().__init__(context)
         self.flow = websocket.WebSocketFlow(context.client, context.server, handshake_flow)
         assert context.server.connected
+        self.client_frame_buffer = []
+        self.server_frame_buffer = []
+        # TODO: get server extensions from handshake flow
+        self.client_conn = wsconn.WSConnection(wsconn.SERVER, wsconn.ConnectionState.OPEN)
+        self.server_conn = wsconn.WSConnection(wsconn.CLIENT, wsconn.ConnectionState.OPEN)
 
     @expect(events.Start)
     def start(self, _) -> commands.TCommandGenerator:
-        yield from ()
+        yield commands.Hook("websocket_start", self.flow)
         self._handle_event = self.relay_messages
 
+    _handle_event = start
+
     @expect(events.DataReceived, events.ConnectionClosed)
     def relay_messages(self, event: events.Event) -> commands.TCommandGenerator:
-        raise NotImplementedError()
+        if isinstance(event, events.DataReceived):
+            from_client = event.connection == self.context.client
+            if from_client:
+                source = self.client_conn
+                other = self.server_conn
+                fb = self.client_frame_buffer
+                send_to = self.context.server
+            else:
+                source = self.server_conn
+                other = self.client_conn
+                fb = self.server_frame_buffer
+                send_to = self.context.client
 
-    _handle_event = start
+            source.receive_bytes(event.data)
+            ws_event = next(source.events())
```

Comment:
Potentially, yes but here's a [list of events](https://github.com/ujjwal96/wsproto/blob/master/wsproto/events.py) that wsproto generates. Since we're not handling the handshake here, only events that should be generated are ConnectionClosed and DataReceived. 
We can have an else block maybe and raise Unexpected event exception?